src/libcore/ptr/unique.rs

```diff
@@ -176,6 +181,7 @@ impl<T: ?Sized> From<&mut T> for Unique<T> {
 impl<T: ?Sized> From<&T> for Unique<T> {
     #[inline]
     fn from(reference: &T) -> Self {
+        // SAFETY: A reference cannot be null
```

Comment:
My understanding is that `Unique` guarantees that its referent is not aliased. This is not guaranteed by `&T`, so I think these `From` impls are unsound. cc @rust-lang/libs 

It's important to state why *all* invariants are upheld when annotating these `unsafe` blocks, not just the one about null pointers.

Reply:
@RalfJung I think mentioned recently that while Unique perhaps *should* provide that guarantee, it does not today.

I agree though that this impl seems very suspicious -- I believe Unique is std-internal today, so we can probably drop this impl?

Reply:
In the docs for `Unique::new_unchecked` , it is only stated that the pointer must be non-null

Reply:
I don’t believe `Unique` has such a language-level guarantee.

What if we just removed it? https://github.com/rust-lang/rust/pull/71530

Reply:
> I don’t believe Unique has such a language-level guarantee.

It doesn't right now mostly because we don't know how to exactly *phrase* the guarantee.

I agree that an instance turning a shared reference into `Unique` sounds wrong, and should be removed.

Reply:
The instance was added in https://github.com/rust-lang/rust/pull/42959. Not sure why noone complained back then.

Reply:
I think this is outside the scope of this pull request, which aims at commenting some unsafe blocks. I'll remove it.